File: modules/swagger-core/src/test/scala/converter/JodaDateTimeConverterTest.scala
package converter

import model._

import com.wordnik.swagger.annotations._
import com.wordnik.swagger.converter._
import com.wordnik.swagger.core.util._
import com.wordnik.swagger.model._

import org.joda.time.DateTime

import scala.collection.mutable.LinkedHashMap
import scala.annotation.meta.field


import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FlatSpec
import org.scalatest.matchers.ShouldMatchers

@RunWith(classOf[JUnitRunner])
class JodaDateTimeConverterTest extends FlatSpec with ShouldMatchers {
  it should "read a generic model" in {
    val models = ModelConverters.readAll(classOf[ModelWithJodaDateTime])
    models.size should be (1) // don't create a Joda DateTime object

    val model = models.head
    val nameProperty = model.properties("name")
    nameProperty.`type` should be ("string")
    nameProperty.position should be (2)
    nameProperty.description should be (Some("name of the model"))

    val dateTimeProperty = model.properties("createdAt")
    dateTimeProperty.`type` should be ("Date")
    dateTimeProperty.position should be (1)
    dateTimeProperty.required should be (true)
    dateTimeProperty.description should be (Some("creation timestamp"))

    println(JsonSerializer.asJson(models.head))
  }
}


case class ModelWithJodaDateTime (
  @(ApiModelProperty @field)(value = "name of the model", position = 2) name: String,
  @(ApiModelProperty @field)(value = "creation timestamp", required = true, position = 1) createdAt: DateTime)

File: modules/swagger-core/src/test/java/io/swagger/v3/core/resolving/JodaDateTimeConverterTest.java
package io.swagger.v3.core.resolving;

import io.swagger.v3.core.converter.ModelConverters;
import io.swagger.v3.oas.models.media.DateTimeSchema;
import io.swagger.v3.oas.models.media.Schema;
import io.swagger.v3.oas.models.media.StringSchema;
import org.joda.time.DateTime;
import org.testng.annotations.Test;

import java.util.Map;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;

public class JodaDateTimeConverterTest {

    @Test
    public void testJodaDateTime() {
        final Map<String, Schema> models = ModelConverters.getInstance().read(ModelWithJodaDateTime.class);
        assertEquals(models.size(), 1); // don't create a Joda DateTime object

        final Schema model = models.get("ModelWithJodaDateTime");

        final Schema dateTimeProperty = (Schema)model.getProperties().get("createdAt");
        assertTrue(dateTimeProperty instanceof DateTimeSchema);
        assertTrue(model.getRequired().contains("createdAt"));
        assertEquals(dateTimeProperty.getDescription(), "creation timestamp");

        final Schema nameProperty = (Schema)model.getProperties().get("name");
        assertTrue(nameProperty instanceof StringSchema);
        assertEquals(nameProperty.getDescription(), "name of the model");
    }

    class ModelWithJodaDateTime{

        @io.swagger.v3.oas.annotations.media.Schema(description = "name of the model")
        public String name;

        @io.swagger.v3.oas.annotations.media.Schema(description = "creation timestamp", required = true)
        public DateTime createdAt;
    }

}
